### ISSUE TITLE
Webcam flag is the wrong way round

### ISSUE THREAD

**Issue by jrudolph** [tuxedocomputers/tuxedo-control-center#36]
It seems in some cases the flag "Webcam on" actually means "Webcam off" and vice versa.

**Comment by jrudolph** [comment]
Setting "on", but hardware flag off:

![Screenshot from 2020-08-03 10-27-53](https://user-images.githubusercontent.com/9868/89162402-4a700780-d574-11ea-9d5b-8b11e627e4c6.png)

Setting "off", but hardware flag on:

![Screenshot from 2020-08-03 10-29-04](https://user-images.githubusercontent.com/9868/89162408-4d6af800-d574-11ea-98ee-8bfca33094db.png)

(That's an old W25CEW, so it might be hardware related but I doubt it)

**Comment by jrudolph** [comment]
Also shows in the log:

```
|Aug  3 10:34:43 mary tccd[11763]: Set webcam status OFF                                                                                                       │
│Aug  3 10:34:43 mary tccd[11763]: Daemon started                                                                                                              │
│Aug  3 10:34:44 mary kernel: [ 4303.473712] usb 1-1.3: new high-speed USB device number 27 using ehci-pci                                                     │
│Aug  3 10:34:44 mary kernel: [ 4303.565707] usb 1-1.3: device descriptor read/64, error -71                                                                   │
│Aug  3 10:34:44 mary kernel: [ 4303.867754] usb 1-1.3: New USB device found, idVendor=5986, idProduct=0401, bcdDevice= 6.06                                   │
│Aug  3 10:34:44 mary kernel: [ 4303.867756] usb 1-1.3: New USB device strings: Mfr=1, Product=2, SerialNumber=0                                               │
│Aug  3 10:34:44 mary kernel: [ 4303.867757] usb 1-1.3: Product: BisonCam, NB Pro                                                                              │
│Aug  3 10:34:44 mary kernel: [ 4303.867759] usb 1-1.3: Manufacturer: Bison                 
```

**Comment by jrudolph** [comment]
It seems every write to `W_WEBCAM_SW` regardless of the argument **toggles** the current state. So, it's not possible to set the status correctly without reading the current state first.

**Comment by jrudolph** [comment]
Here I'm continuously writing `0` into `W_WEBCAM_SW` and read `R_WEBCAM_SW` afterwards, it shows that the flag toggles all the time:

```
johannes@mary ~os/_2020/ioctl (git)-[master|…] λ echo -n "\x00\x00\x00\x00\x00\x00\x00\x00" | sudo ./ioctl /dev/tuxedo_cc_wmi 0x4008ed12 && sudo ./ioctl /dev/tuxedo_cc_wmi 0x8008ec13|hexdump -C
Decoded values: ioctl=0x4008ed12, direction=W, arg size=8 bytes, device number=0xed, function number=18
Used values: ioctl=0x4008ed12, direction=W, arg size=8 bytes, device number=0xed, function number=18
Returned 0
Decoded values: ioctl=0x8008ec13, direction=R, arg size=8 bytes, device number=0xec, function number=19
Used values: ioctl=0x8008ec13, direction=R, arg size=8 bytes, device number=0xec, function number=19
Returned 0
00000000  01 00 00 00 00 00 00 00                           |........|
00000008
johannes@mary ~os/_2020/ioctl (git)-[master|…] λ echo -n "\x00\x00\x00\x00\x00\x00\x00\x00" | sudo ./ioctl /dev/tuxedo_cc_wmi 0x4008ed12 && sudo ./ioctl /dev/tuxedo_cc_wmi 0x8008ec13|hexdump -C
Decoded values: ioctl=0x4008ed12, direction=W, arg size=8 bytes, device number=0xed, function number=18
Used values: ioctl=0x4008ed12, direction=W, arg size=8 bytes, device number=0xed, function number=18
Returned 0
Decoded values: ioctl=0x8008ec13, direction=R, arg size=8 bytes, device number=0xec, function number=19
Used values: ioctl=0x8008ec13, direction=R, arg size=8 bytes, device number=0xec, function number=19
Returned 0
00000000  00 00 00 00 00 00 00 00                           |........|
00000008
johannes@mary ~os/_2020/ioctl (git)-[master|…] λ echo -n "\x00\x00\x00\x00\x00\x00\x00\x00" | sudo ./ioctl /dev/tuxedo_cc_wmi 0x4008ed12 && sudo ./ioctl /dev/tuxedo_cc_wmi 0x8008ec13|hexdump -C
Decoded values: ioctl=0x4008ed12, direction=W, arg size=8 bytes, device number=0xed, function number=18
Used values: ioctl=0x4008ed12, direction=W, arg size=8 bytes, device number=0xed, function number=18
Returned 0
Decoded values: ioctl=0x8008ec13, direction=R, arg size=8 bytes, device number=0xec, function number=19
Used values: ioctl=0x8008ec13, direction=R, arg size=8 bytes, device number=0xec, function number=19
Returned 0
00000000  01 00 00 00 00 00 00 00                           |........|
00000008
```

Could of course be an artifact of that old W25CEW implementation. You should have more clevos to test...

**Comment by tuxedoxt** [comment]
Hello, W25CEW must be rather old indeed, unfortunately I do not have a device of this type at hand.

The behaviour could be an interface bug or old functionality. Either way, if it works as you suggest and it toggles the value no matter if writing a 1 or a 0, it should be an easy workaround checking the value before setting it. (As long as the read works reliably on all devices that is..)

Would you mind sharing the `/sys/firmware/acpi/tables/DSDT` of the device?

**Comment by jrudolph** [comment]
Thanks for the quick response.

Here's the DSDT:

[dsdt.dat.gz](https://github.com/tuxedocomputers/tuxedo-control-center/files/5020786/dsdt.dat.gz)

If the problem is really only related to my device, I don't expect you to fix it.

> Hello, W25CEW must be rather old indeed, unfortunately I do not have a device of this type at hand.

Yep it's a Tuxedo BX1501 from 2012... Still going strong thanks to the 3840QM. And now finally after all this time, I found out that the fan control of tuxedo-control-center fixes the main problem with that laptop which was an annoyingly broken BIOS fan control so that fans were running much too loud most of the time. So, no hard feelings if that cannot be fixed for just that device :) (Also I ordered a Pulse 15 so this notebook might be retired soon enough...)

**Comment by tuxedoxt** [comment]
Thanks, nice that it has held up.

I had a quick look. It seems, indeed, like the previous interface was just toggle, it does not take an argument into account. The get looks similar though. Therefore only a small check should be necessary for our interface to support both versions.

I implemented it in the interface tuxedo-cc-wmi. If it works I will include it in the next release. Would you be able to test it?

To manually build and temporarily load the module (needs make, gcc, kernel headers)

```
git clone https://github.com/tuxedocomputers/tuxedo-cc-wmi.git
cd tuxedo-cc-wmi
make
sudo rmmod tuxedo_cc_wmi; sudo insmod src/tuxedo_cc_wmi.ko
```

**Comment by jrudolph** [comment]
Looking good! Thanks a lot.

**Comment by tuxedoxt** [comment]
Closing, fix included in tuxedo-cc-wmi 0.1.5
